src/rpc/pd/client.rs

```diff
@@ -0,0 +1,218 @@
+        })
+    }
+
+    fn execute<Resp, Executor, RpcFuture>(
+        &self,
+        mut context: PdRequestContext<Executor>,
+    ) -> impl Future<Item = Resp, Error = Error>
+    where
+        Resp: PdResponse + Send + fmt::Debug + 'static,
+        RpcFuture: Future<Item = Resp, Error = ::grpcio::Error> + Send + 'static,
+        Executor: FnMut(&RpcClient, CallOption) -> ::grpcio::Result<RpcFuture> + Send + 'static,
+    {
+        let timeout = self.timeout;
+        let mut executor = context.executor();
+        let wrapper = move |cli: &RwLock<LeaderClient>| {
+            let option = CallOption::default().timeout(timeout);
+            let cli = &cli.rl().client;
+            executor(cli, option)
+                .unwrap()
+                .map_err(Error::Grpc)
+                .and_then(|r| {
+                    {
+                        let header = r.header();
+                        if header.has_error() {
+                            return Err(internal_err!(header.get_error().get_message()));
+                        }
+                    }
+                    Ok(r)
+                })
+        };
+        Request::new(
+            wrapper,
+            Arc::clone(&self.leader),
+            LeaderClient::reconnect,
+            LEADER_CHANGE_RETRY,
+        )
+        .execute()
+        .then(move |r| context.done(r))
+    }
+
+    pub fn get_all_stores(&self) -> impl Future<Item = Vec<Store>, Error = Error> {
+        let req = pd_request!(self.cluster_id, pdpb::GetAllStoresRequest);
+
+        self.execute(request_context(
+            "get_all_stores",
+            move |cli: &RpcClient, opt: _| cli.get_all_stores_async_opt(&req, opt),
+        ))
+        .map(|mut resp| resp.take_stores().into_iter().map(Into::into).collect())
+    }
+
+    pub fn get_store(&self, store_id: StoreId) -> impl Future<Item = Store, Error = Error> {
+        let mut req = pd_request!(self.cluster_id, pdpb::GetStoreRequest);
+        req.set_store_id(store_id);
+
+        self.execute(request_context(
+            "get_store",
+            move |cli: &RpcClient, opt: _| cli.get_store_async_opt(&req, opt),
+        ))
+        .map(|mut resp| resp.take_store().into())
+    }
+
+    pub fn get_region(&self, key: &[u8]) -> impl Future<Item = Region, Error = Error> {
+        self.get_region_and_leader(key)
+            .map(|x| Region::new(x.0, x.1))
+    }
+
+    pub fn get_region_by_id(&self, id: RegionId) -> impl Future<Item = Region, Error = Error> {
+        self.get_region_and_leader_by_id(id)
+            .map(|x| Region::new(x.0, x.1))
+    }
+
+    pub fn get_ts(&self) -> impl Future<Item = PdTimestamp, Error = Error> {
+        self.leader.wl().get_ts()
+    }
+}
+
+impl fmt::Debug for PdClient {
+    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
+        fmt.debug_struct("PdClient")
+            .field("cluster_id", &self.cluster_id)
+            .field("leader", &self.get_leader())
```

Comment:
What about `timeout`? Would it be useful do you think?

Reply:
Let's add it here.